cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/service/subnetselector/SubnetSelectorService.java

```diff
@@ -0,0 +1,78 @@
+package com.sequenceiq.cloudbreak.cloud.aws.service.subnetselector;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.cloud.model.CloudSubnet;
+
+@Service
+class SubnetSelectorService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SubnetSelectorService.class);
+
+    Map<String, CloudSubnet> collectOnePrivateSubnetPerAz(List<CloudSubnet> subnetMetas, int max) {
+        Map<String, CloudSubnet> subnetsPerAz = new HashMap<>();
+        Iterator<CloudSubnet> subnetIterator = subnetMetas.iterator();
+        while (subnetsPerAz.size() < max && subnetIterator.hasNext()) {
+            CloudSubnet nextSubnet = subnetIterator.next();
+            if (nextSubnet.isPrivateSubnet()) {
+                subnetsPerAz.putIfAbsent(nextSubnet.getAvailabilityZone(), nextSubnet);
+            }
+        }
+        LOGGER.debug("Private subnets per AZ: {}", subnetsPerAz.values().stream().map(CloudSubnet::getId).collect(Collectors.joining(",")));
+        return subnetsPerAz;
+    }
+
+    Optional<CloudSubnet> getOnePrivateSubnet(List<CloudSubnet> subnetMetas) {
+        Optional<CloudSubnet> foundCloudSubnet = subnetMetas.stream()
+                .filter(CloudSubnet::isPrivateSubnet)
+                .findFirst();
+        LOGGER.debug("Found private subnet: {}", foundCloudSubnet.map(CloudSubnet::getId).orElse("Not found"));
+        return foundCloudSubnet;
+    }
+
+    Map<String, CloudSubnet> collectOnePublicSubnetPerAz(List<CloudSubnet> subnetMetas, int max) {
+        Map<String, CloudSubnet> subnetsPerAz = new HashMap<>();
+        Iterator<CloudSubnet> subnetIterator = subnetMetas.iterator();
+        while (subnetsPerAz.size() < max && subnetIterator.hasNext()) {
+            CloudSubnet nextSubnet = subnetIterator.next();
+            if (isUsablePublicSubnet(nextSubnet)) {
+                subnetsPerAz.putIfAbsent(nextSubnet.getAvailabilityZone(), nextSubnet);
+            }
+        }
```

Comment:
it's similar to the private one, could be replaced with stream.

Reply:
Thanks, did it.

Reply:
Finally I kept the original implementation.